annotators/bot_emotion_classifier/server.py

```diff
@@ -0,0 +1,425 @@
+
+default_mood = [pleasure, arousal, dominance]
+
+
+# get bot emotion using user emotion and user utterance
+def get_bot_emotion(sent, emotion, sentiment):
+    if emotion == "neutral":
+        bot_emotion = "neutral"
+    elif emotion == "surprise":
+        if sentiment == "negative":
+            type_num = get_dsript_type(sent, neg_type_expl)
+            bot_emotion = neg_reactions[type_num][emotion]
+        elif sentiment == "positive":
+            type_num = get_dsript_type(sent, pos_type_expl)
+            bot_emotion = pos_reactions[type_num][emotion]
+        else:
+            bot_emotion = "neutral"
+    elif emotion in negative_emotions:
+        type_num = get_dsript_type(sent, neg_type_expl)
+        bot_emotion = neg_reactions[type_num][emotion]
+    else:
+        type_num = get_dsript_type(sent, pos_type_expl)
+        bot_emotion = pos_reactions[type_num][emotion]
+    return bot_emotion
+
+
+# the rate of mood decrease
+def get_dim_decay(default_dim, curr_dim):
+    p_dif = abs(default_dim - curr_dim)
+    if p_dif == 0:
+        p_dif = 0.00001
+    decay = 1 / p_dif * 0.01
+    if decay > p_dif:
+        decay = p_dif
+    if default_dim < curr_dim:
+        decay = -decay
+    return decay
+
+
+def get_decay(default_mood, curr_mood):
+    decay = []
+    for i in range(len(default_mood)):
+        decay.append(get_dim_decay(default_mood[i], curr_mood[i]))
+    return decay
+
+
+# comparison of old and new mood
+def check_same_mood(curr_mood, new_mood):
+    print(curr_mood, new_mood)
+    for i in range(len(curr_mood)):
+        if curr_mood[i] * new_mood[i] < 0:
+            return False
+    return True
+
+
+# new mood calculation using difference btw current mood and current emotion
+def get_new_mood(default_mood, curr_mood, bot_emotion):
+    if check_same_mood(pad_emotions[bot_emotion], curr_mood):
+        decay = [0, 0, 0]
+    else:
+        decay = get_decay(default_mood, curr_mood)
+    vec = [pad_emotions[bot_emotion][i] - curr_mood[i] for i in range(len(curr_mood))]
+    new_mood = [0.5 * curr_mood[i] + 0.5 * vec[i] + decay[i] for i in range(len(curr_mood))]
+    new_mood_reg = [1 if dim > 1 else dim for dim in new_mood]
+    new_mood_reg = [-1 if dim < -1 else dim for dim in new_mood_reg]
+
+    # dim_symbols = [str(int(dim / abs(dim))) if dim != 0 else '-1' for dim in new_mood_reg]
+    # octant = ''.join(dim_symbols)
+    # print('New mood:', pad_moods[octant])
+    return new_mood_reg
+
+
+def get_mood_label(bot_mood):
+    octant = ""
+    for dim in bot_mood:
+        if dim == 0:
+            octant += "0"
+        elif dim > 0:
+            octant += "1"
+        else:
+            octant += "-1"
+
+    return pad_moods[octant]
+
+
+@app.route("/model", methods=["POST"])
+def respond():
+    sentences = request.json.get("sentences", [])
+    annotated_utterances = request.json.get("annotated_utterances", [])
+    bot_moods = request.json.get("bot_mood", [])
+
+    for i in range(len(sentences)):
```

Comment:
just writing to show you how it is more beautiful to write these 4 lines of code in one:
```
for sentence, annotated_utterance, bot_mood in zip(sentences, annotated_utterances, bot_moods):
```

Reply:
I don't know why but sometimes `zip` confuses me in code of other people, and I find `for i` easier to understand, but I will try to use `zip` instead because it's truly more beautiful :)